Only apply default margin from global styles if root body does not have margin set in theme.json

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

Note there is now also an alternative approach to this in https://github.com/WordPress/gutenberg/pull/36960.

In https://github.com/WordPress/gutenberg/issues/36147 we learn that the Theme JSON class automatically applies a `margin: 0` to the `body` element. Unfortunately it does so in a way that means if you use `theme.json` to define a margin on the body element, the default `margin: 0` will still take precedence.

For example applying this to `theme.json` will still leave you with the _default_ `margin: 0` on the `body`, rather than the _expected_ value of `margin: 0 5rem`:

```
{
	"styles": {
		"spacing": {
			"margin": "0 5rem",
			"padding": "0"
		},
	}
}
```

This PR fixes this by only applying the default `margin: 0` if the root spacing does not have a margin declaration.

Closes https://github.com/WordPress/gutenberg/issues/36147

## Question

Need to consider what happens how to handle things if the `spacing -> margin` is only set on a single axis (e.g. `margin-left` or `margin-bottom`.

An alternative implementation might be to simply allow the `margin: 0` to be set but to ensure that any margin declarations coming from `theme.json` are applied _afterwards_ rather than before as they are currently. This would allow the `theme.json` declarations to take precedence.

**Update**: I have this approach as a PR in https://github.com/WordPress/gutenberg/pull/36960.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

⚠️ ⚠️ ⚠️ ⚠️ ~You cannot currently test this because the `class-wp-theme-json-gutenberg.php` file is not loaded when WP 5.9 is active. Rather the Core version of the class is used. This PR now includes changes from https://github.com/WordPress/gutenberg/pull/36957 which allow you to test this PR. They will be removed prior to merge~ You will need to use WordPress **5.8** (that is five point **EIGHT** _not_ NINE) to test this PR. See https://github.com/WordPress/gutenberg/pull/36957#issuecomment-981759580 for reasons. ⚠️ ⚠️ ⚠️ 

@oandregal is working on a fix for this at which point you will be able to follow the testing instructions below:

0. Make sure you are running WP 5.8!
1. Install and activate TT1 blocks (or block Theme or your choice). You should make this theme local in order that you can edit the `theme.json` of the Theme.
2. Load the front of the site. Inspect the `body` element and notice the `margin: 0` applied to the `body` element.
3. Now edit `theme.json` and add the following under the `styles` section:
```
{
	"spacing": {
		"margin": "0 5rem",
		"padding": "0"
	},
}
```
4. Reload the front of the site and inspect the `body`. Notice how the `margin: 0` is no longer applied and instead the applied style of the `body` is `margin: 0 5rem`.

⚠️  **Watch out**: [TT1 blocks applies `margin: 0` to `body` in its `style.css` file](https://github.com/WordPress/theme-experiments/blob/dff7d86e56f3657b492fdeae3e7e6765a00f22c0/tt1-blocks/style.css#L24-L26). Don't get confused when verifying the results of this PR which have nothing to do with the manually created `style.css` file.




## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave commented:
cc @overclokk as you raised the original Issue.

getdave commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1059,7 +1059,11 @@ private function get_block_classes( $style_nodes ) {
-				$block_rules .= 'body { margin: 0; }';
+				// If the root body node has margin set in its spacing property
+				// then avoid setting a default margin.
+				if ( ! isset( $node['spacing']['margin'] ) ) {
+					$block_rules .= 'body { margin: 0; }';
+				}
```
This is the changeset to focus on in this PR

overclokk replied:
I was thinking why not adding that value as default in theme.json in core, so the theme can override that vales without an if statement

oandregal replied:
Relevant to this fix is the original PR by @youknowriad and this comment https://github.com/WordPress/gutenberg/pull/35421#discussion_r724010968

cc @ntsekouras

ntsekouras replied:
>I considered whether this should be applied in the default theme.json instead but decided against because while right now the default theme.json styles applies to body, I'm wondering if it's better if they apply to .wp-site-blocks because that's the actual container of everything (imagine defining a layout style at the root level, you want it to apply there and not on body)

@youknowriad's comment makes sense but I'm wondering if changing this now would cause some themes to break. Should we handle `margin` now in `body` and consider a change like this after 5.9 🤔 ?

Setting a default in theme.json sounds good, but we also need to handle both notations `margin: 10px` and `margin:{top: 10px}`.

--cc @jasmussen @kjellr

getdave replied:
My personal opinion is that this PR is a _bug fix_ and we should strive to keep it as such.

I understand there are other things to explore, namely:

1. Making this part of default `theme.json` as suggested in https://github.com/WordPress/gutenberg/pull/36958#discussion_r758487621.
2. Applying the `margin: 0` to the `.wp-site-blocks` element as per https://github.com/WordPress/gutenberg/pull/35421#discussion_r724010968.

...however the aim of this PR is to fix the bug ready for WP 5.9 Beta 2 (given that Beta 1 deadline has passed) and it seems to me - given we're in a "bug fix only" period -  we should therefore keep the scope of fixes limited.

Personally I would prefer we pursued https://github.com/WordPress/gutenberg/pull/36960 as this option retains the current `margin: 0` on the `body` which is what Themes have come to expect, but it also allows Themes to overide that setting via `theme.json`. That seems to cover the use case stated in the original Issue #36147, namely:

> If I add a margin property to the theme.json at the top level styles the style generated is overwritten by a style added by the gutenberg plugin.

Given [the overhead of raising and testing Global Styles PR](https://github.com/WordPress/gutenberg/pull/36957#issuecomment-981759580)s (PRs in both Gutenberg and Core) it would be great if we could decide on the preferred approach in order that we can get this issues resolved with maximum efficiency. I appreciate your understanding.

The decisions to be made are:

1. Do we agreed that we should proceed with this "limited" fix scope (i.e. not pursuing other explorations)?
2. Which PR should we pursue?:
    - #36958
    - ...OR... 
    - #36960

Thanks all 🙇

ntsekouras replied:
>Do we agreed that we should proceed with this "limited" fix scope (i.e. not pursuing other explorations)?

I personally agree with this.

>Which PR should be pursue 

I also prefer the selector ordering PR (#36960 )

jasmussen replied:
> comment makes sense but I'm wondering if changing this now would cause some themes to break. Should we handle margin now in body and consider a change like this after 5.9 🤔 ?

As a high level principle, I like the idea of serving out good defaults but giving full control to themers to trivially style everything with theme.json, style.css or otherwise.

In this case it sounds like the risk is themes assuming `body { margin: 0; }` to be present, and this PR then removing it, is that correct?

In that case, have we considered adding this rule to create a soft default? 

```
:where(body) {
	margin: 0;
}
```

In theory that should output the existing rule but be trivially overridden by theme.json.

ntsekouras replied:
>In this case it sounds like the risk is themes assuming body { margin: 0; } to be present, and this PR then removing it, is that correct?

I think it's not only about `margin`, but all the other 'root' styles to be applied to `.wp-site-blocks` and not `body`.

getdave replied:
@jasmussen An alternative fix is https://github.com/WordPress/gutenberg/pull/36960

getdave replied:
See https://github.com/WordPress/gutenberg/pull/36958#issuecomment-982549243

getdave commented:
In light of https://github.com/WordPress/gutenberg/pull/36957#issuecomment-981759580 I've removed the fix I cherrypicked from https://github.com/WordPress/gutenberg/pull/36957.

I also rebased to pick up recent changes in `trunk`.

You now need to test this PR against WordPress **5.8**.

getdave committed:
Only reset margin if no margin on root in theme.json

overclokk commented:
Still I don't understand what is the issue of using the theme.json in core as default values instead of the if statement.

getdave commented:
I'm going to close this out to direct attention towards https://github.com/WordPress/gutenberg/pull/36960.

If folks decide that's not the best option we can always reopen this one.

getdave commented:
> Still I don't understand what is the issue of using the theme.json in core as default values instead of the if statement.

@overclokk ~Let me spin up a PR as an alternative approach. But I think the problem will be that if the Theme uses the object notation for `margin` (e.g. `margin-left: 0`) then that will override the entirety of the `margin: 0` reset in Core.~ I've listed several problems as I see them.~

~Another possible reason for including the `if()` statement is to make it clear that this is a CSS _reset_ rather than some abstract default provided by Core.~

I've spun up https://github.com/WordPress/gutenberg/pull/36996 as an alternative which includes your suggestion. I've listed several concerns I have.
